### PR TITLE
Allow by-ref Extension methods on value types 

### PR DESCRIPTION
**Pull request by pakrym**
Reopening https://github.com/dotnet/roslyn/pull/15535 targeting 'features/moreRef'

### CODE REVIEW

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 1]
```diff
@@ -4984,4 +4987,7 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="ERR_PossibleBadNegCast" xml:space="preserve">
     <value>To cast a negative value, you must enclose the value in parentheses.</value>
   </data>
+  <data name="ERR_RefExtensionMethodOnNonValueType" xml:space="preserve">
+    <value>A ref extension methods can only be defined on value types.</value>
```
Comment (davidfowl) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: method*

**Comment by jamesqo** [comment]
Why was this closed? :cry:

**Comment by jaredpar** [comment]
@jamesqo there is a bigger effort here around `ref readonly` that includes adding `ref` and `ref readonly` extension methods.  

https://github.com/dotnet/csharplang/blob/master/proposals/readonly-ref.md#refin-extension-methods

This PR was closed because the code is going to flow in through that branch vs. this PR.  In the context of the bigger work extension methods end up get done last.  Once we get to extension method section we're going to merge this code into the [ref-readonly](https://github.com/dotnet/roslyn/tree/features/readonly-ref) branch.

**Comment by jamesqo** [comment]
@jaredpar Ah, okay. I thought this feature was being abandoned-- glad to see that it isn't.

**Pull request closed without merging**